### ISSUE TITLE
cast/forge: contract types not preserved when generating interface

### ISSUE THREAD

**Issue by mds1** [foundry-rs/foundry#8837]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (143abd6 2024-09-04T00:24:41.963834000Z)

### What command(s) is the bug in?

cast interface / forge inspect

### Operating System

None

### Describe the bug

To reproduce, `forge init` a new project and add this contract:

```solidity
contract Foo {
  function bar(Counter x) public view returns (Counter y) {
    return x;
  }
}
```

When generating an interface, the function inputs and return type both are converted to `address`, which strips information. You can see the `Counter` type is present in the ABI of Foo:

```json
"abi": [
  {
    "type": "function",
    "name": "bar",
    "inputs": [
      {
        "name": "x",
        "type": "address",
        "internalType": "contract Counter"
      }
    ],
    "outputs": [
      {
        "name": "y",
        "type": "address",
        "internalType": "contract Counter"
      }
    ],
    "stateMutability": "view"
  }
],
```

But when generating interfaces:

```sh
$ forge inspect Foo abi --pretty
interface Foo {
    function bar(address x) external view returns (address y);
}

$ cast interface Foo
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.4;

interface Foo {
    function bar(address x) external view returns (address y);
}
```

I can see the rationale for this being that forge might not know what path to use for the `Counter` import. However, even if no `import` statement is provided, I would still prefer a stronger version of interface generation that preserves the internal `Counter` type. Perhaps this should be behind a `--preserve-internal-types` flag

cc @smartcontracts

**Comment by DaniPopes** [comment]
There is no information about `Counter` in `Foo`, so I don't see the point in preserving this since we would have to generate an empty `Counter` interface; in any case they're ABI-equivalent

**Comment by mds1** [comment]
Right, forge would generate an incomplete interface, and the user would have to fix it by importing Counter from the proper path. The value here is that having Counter in the interface is a stronger type than address, and currently the user must manually fix the interface to preserve the stronger typing. This manual fix is tedious and error prone for large interfaces
